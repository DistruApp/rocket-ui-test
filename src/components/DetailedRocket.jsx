import React from 'react';
import PropTypes from 'prop-types';

const DetailedRocket = ({ rocket }) => (
  <div>
    <div> Rocket ID: { rocket.id } </div>
    <div> Rocket Cost per launch: { rocket.cost_per_launch } </div>
    <div> Rocket Description: { rocket.description } </div>
  </div>
);
DetailedRocket.propTypes = {
  rocket: PropTypes.shape(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      cost_per_launch: PropTypes.number.isRequired,
      description: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired
};

export default DetailedRocket
