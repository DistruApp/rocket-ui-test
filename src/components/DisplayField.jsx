import React from 'react';
import PropTypes from 'prop-types';

const DisplayField = ({ label, value }) => <p style={{ padding: '4px 0' }}><strong>{label}:</strong>&nbsp;{value}</p>

DisplayField.propTypes = {
  label: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ])
}

export default DisplayField;