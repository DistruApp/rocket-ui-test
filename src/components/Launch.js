import React from 'react';

const Launch = (props) => {
	const { launch } = props;

	return (
		<li>
			<h2> {launch.mission_name} </h2>
			<div> Flight Number: {launch.flight_number} </div>
		</li>
	);
};

export default Launch;
