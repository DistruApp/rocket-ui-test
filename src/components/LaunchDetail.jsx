import React, {useContext} from 'react';
import { store } from "../stores/store";

const Launch = ({ launch }) => {
    const { dispatch, state } = useContext(store)

    const rocket = state.rockets.filter(rocket => rocket.id === launch.rocket)[0]

    return (
        <>
            <div className="rocket-detail">
                    {
                        launch.rocket && rocket &&
                            <>
                                <div>
                                    <h5>
                                        Rocket Name
                                    </h5>
                                    <span>
                                        { rocket.name }
                                    </span>
                                    <h5>
                                        Rocket ID
                                    </h5>
                                    <span>
                                        { launch.rocket }
                                    </span>
                                </div>
                                <div>
                                    <h5>
                                        Cost per Launch
                                    </h5>
                                    <span>
                                        { `$${rocket.cost_per_launch.toLocaleString()}` }
                                    </span>
                                </div>
                            </>
                    }
            </div>
            <div className="launch-detail">
                {
                    launch.details &&
                        <div>
                            <h5>
                                Details
                            </h5>
                            <span>
                                { launch.details }
                            </span>
                        </div>
                }
            </div>
        </>
    );
}

export default Launch;
