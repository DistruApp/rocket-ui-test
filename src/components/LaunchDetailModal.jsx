import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';

const LaunchDetailModal = ({ open, handleClose, currentLaunch = {} }) => {
  const { mission_name } = currentLaunch;
  return <Modal isOpen={open}>
    <div className="launch-detail-modal">
      <h2>{mission_name}</h2>

    </div>
  </Modal>
}

LaunchDetailModal.PropTypes = {
  handleClose: PropTypes.func,

}

export default LaunchDetailModal;