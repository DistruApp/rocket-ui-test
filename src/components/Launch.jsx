import React from 'react';
import { connect } from "react-redux";

import { shouldFetchRocket, fetchRocket } from "../actions/Rockets";
import { getRocket } from "../queries/Rockets";
import DetailedRocket from "./DetailedRocket";

const Launch = ({ launch, active, rocket, dispatch }) => {
  const rocketId = launch.rocket;
  // Fetching in this component so that when fetch occurs, only this component
  // updates and not all the other launches
  if (active && !rocket) fetchRocket(dispatch, rocketId);

  return (
    <div>
      <h2> { launch.name } </h2>
      <div> Flight Number: { launch.flight_number } </div>
      { active && rocket && <DetailedRocket rocket={rocket} /> }
    </div>
  );
}

const mapStateToProps = ({ rocketCollection }, { launch: { rocket: rocketId } }) => {
  const rocket = getRocket(rocketCollection, rocketId);
  return { rocket }
};
const mapDispatchToProps = dispatch => ({ dispatch });

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Launch);
