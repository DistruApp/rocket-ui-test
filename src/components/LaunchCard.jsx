/* eslint-disable camelcase */
import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import clsx from 'clsx';
import DisplayField from './DisplayField'

const LaunchCard = ({ launch = {}, launchOnClick, selected }) => {
  const { mission_name, flight_number, launch_success } = launch;
  const missionPatch = _.get(launch, 'links.mission_patch_small');


  return <li className={clsx('launch-card', { selected })} onClick={() => launchOnClick(launch)} role="button">
    <div className="launch-card-image">
      {missionPatch ? <img src={missionPatch} alt={`${mission_name}`} /> : <div className="mission-patch-not-found">MISSION PATCH NOT AVAILABLE</div>}
    </div>
    <div className="launch-card-detail">
      <div>
        <h3> {mission_name} </h3>
      </div>
      <DisplayField label="Flight Number" value={flight_number} />
      <div>
        <h5 className={clsx({ 'text-success': launch_success, 'text-error': !launch_success })} >MISSION {launch_success ? 'SUCCESS' : 'FAILURE'}</h5>
      </div>
    </div>
  </li>
}

LaunchCard.propTypes = {
  launch: PropTypes.object.isRequired,
  launchOnClick: PropTypes.func.isRequired,
  selected: PropTypes.bool.isRequired,
}

export default LaunchCard;
