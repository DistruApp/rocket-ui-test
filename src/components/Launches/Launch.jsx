import React, { useContext } from 'react';
import PropTypes from "prop-types";
import { store } from "../../stores/store";
import LaunchDetail from "./LaunchDetail";

const Launch = ({ launch }) => {
    const { dispatch, state } = useContext(store);

    const showLaunchDetail = state.ui.showLaunchDetail.show && state.ui.showLaunchDetail.id === launch.id;

    const toggleShowDetail = (launchId) => {
        dispatch({type: 'set-show-launch-detail', payload: {show: !showLaunchDetail, id: launchId}});
    }

    return (
        <div className="launch-container"
            onClick={() => toggleShowDetail(launch.id)}
        >
            <h2>
                { launch.name }
            </h2>
            <div className="launch-overview">
                <div>
                    <h5>Flight Number</h5>
                    <span>
                        { launch.flight_number }
                    </span>
                </div>
                <div>
                    <h5>
                        Success
                    </h5>
                    <span>
                        { launch.success ? 'Yes' : 'No' }
                    </span>
                </div>
            </div>
            {
                showLaunchDetail && <LaunchDetail launch={launch} />
            }
        </div>
    );
}

Launch.propTypes = {
    launch: PropTypes.shape({
        id: PropTypes.string.isRequired,
        flight_number: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        success: PropTypes.bool.isRequired,
    }).isRequired
}

export default Launch;
