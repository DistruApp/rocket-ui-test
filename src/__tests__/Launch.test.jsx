import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as redux from 'react-redux';
import { Launch } from '../components/Launch';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}));

const exampleLaunchDef = {
  rocket: {
    rocket_id: '1'
  },
  _id: '1',
  mission_name: 'Test Mission Name',
  flight_number: 'Test Flight Number'
};

configure({ adapter: new Adapter() });

describe('Launch component test', () => {
  it('renders Launch component without crashing', () => {
    const component = mount(<Launch launch={exampleLaunchDef} onClick={() => {}} isOpen={false} />);
    expect(component.getElements()).toMatchSnapshot();
  });
  it('tests Launch component behaviour', () => {
    const mockOnClick = jest.fn();
    const useDispatchSpy = jest.spyOn(redux, 'useDispatch');
    const mockDispatchFn = jest.fn();
    useDispatchSpy.mockReturnValue(mockDispatchFn);
    const component = mount(
      <Launch launch={exampleLaunchDef} onClick={mockOnClick} isOpen={false} />
    );
    expect(component.exists('table')).toBeFalsy();
    component.find('button').simulate('click');
    expect(mockOnClick.mock.calls.length).toEqual(1);
    expect(component.find('table')).toBeTruthy();
    component.find('button[type="button"]').simulate('click');
    expect(mockOnClick.mock.calls.length).toEqual(2);
    expect(component.exists('table')).toBeFalsy();
    useDispatchSpy.mockClear();
  });
});
