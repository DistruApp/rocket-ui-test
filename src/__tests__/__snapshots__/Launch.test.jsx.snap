// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launch component test renders Launch component without crashing 1`] = `
Array [
  <Launch
    isOpen={false}
    launch={
      Object {
        "_id": "1",
        "flight_number": "Test Flight Number",
        "mission_name": "Test Mission Name",
        "rocket": Object {
          "rocket_id": "1",
        },
      }
    }
    onClick={[Function]}
  />,
]
`;
